### ISSUE TITLE
ARRAYS_QC_BLACKLISTING.MODIFIED_AT field (and possibly other timestamp fields) appear incorrectly set in local time

### ISSUE THREAD

**Issue by kbergin** [HumanCellAtlas/secondary-analysis#676]
2 data incidents 

* [https://broadinstitute.atlassian.net/browse/PO-18234](https://broadinstitute.atlassian.net/browse/PO-18234) and 
* [https://broadinstitute.atlassian.net/browse/PO-18311](https://broadinstitute.atlassian.net/browse/PO-18311)

and 1 test 

* [https://gpinfojira.broadinstitute.org/jira/browse/RPT-5504](https://gpinfojira.broadinstitute.org/jira/browse/RPT-5504)

suggest that **ArraysManualBlacklistUpdate** command from picard-private.jar is recording local time in MODIFIED_AT field which is broadly used as a UTC field. Effectively this means that upon arrival in the DB new records appear as timestamped 4h ago (in UTC terms) have no chance to catch the AnalyticsETL train.

It seems to me that the tool is reading current time from JVM which is susceptible to specific node settings and worse timezone-setting may vary from one node to another.
(We are seeing similar TimeZone problems with METRICS tables in SEQPROD db.)

Rather than tweaking the Java code, it may be much easier and more solid if timestamping is delegated to the host database: just have a "BEFORE INSERT or UPDATE" trigger set MODIFIED_AT field with SYSDATE(). This approach features:

* easy - 3 lines of SQL to install the trigger in DB
* reliable - no matter how buggy client-apps are, MODIFIED_AT would be reliable set every single time
* no need to fix the Java code
* gaining popularity - same approach is being used here [https://broadinstitute.atlassian.net/browse/GH-229](https://broadinstitute.atlassian.net/browse/GH-229)

It's reasonable to assume that other tables:

* ARRAYS_QC
* ARRAYS_QC_BLACKLISTING
* ARRAYS_QC_CONTROL_CODES
* ARRAYS_QC_FINGERPRINT
* ARRAYS_QC_GT_CONCORDANCE
* CHIP_WELL_BARCODE_INDEX

are also exposed to this problem. Would make a lot of sense to address all of them at once.
Would be happy to provide assistance with testing or anything else.
Help appreciated \!
Nasko

┆Issue is synchronized with this [Jira Story](https://broadinstitute.atlassian.net/browse/GH-317)

**Comment by kbergin** [comment]
➤ Gina Vicente commented:

Hi [~atanas]  Thank you for creating this ticket.  Is it possible to prioritize this please?  The Infinium team has a few other tickets that may need whitelist/blacklist that we will be requesting this week as well.  

Another ticket is: +PO-18195+ ( https://broadinstitute.atlassian.net/browse/PO-18195 )

cc: [~babchuck]  [~kflowers]

**Comment by kbergin** [comment]
➤ Sarah Babchuck commented:

Hi [~atanas] ,

I’m also here to advocate for prioritization. A PDO with a funding deadline this week is currently held up by this issue. Anything we can do to get this moving is much appreciated.

Thanks so much\!

Sarah

**Comment by kbergin** [comment]
➤ Susanna Hamilton commented:

This happened again today. Missing data in the datamarts. Can this fix please be prioritized?

**Comment by kbergin** [comment]
➤ Gina Vicente commented:

Hi [~jkaneria]  [~hensonc]  [~thompson] [~suhamil] [~desmet] [~samn] 

I want to bring attention to this ticket because it is related to the conversation in dgplanning this morning. Infinium has several PO tickets related to this. Thanks\!

**Comment by kbergin** [comment]
➤ Atanas Mihalev commented:

While waiting for the fix:
Running following script every time records are updated/inserted would make sure MODIFIED_AT timestamp is properly populated and new data catch the ETL train. 
Just provide appropriate list of specific (CHIP_WELL_BARCODE, ANALYSIS_VERSION) pairs in the WHERE clause.
ETL trains are coming every 15mins. Make sure all changes (incl. calling the script) are done before 15min-period close (at 00, 15, 30, 45 minutes each hour)

START TRANSACTION;
UPDATE ARRAYS_QC_BLACKLISTING set modified_at = sysdate()
WHERE (CHIP_WELL_BARCODE, ANALYSIS_VERSION) IN (
  ('chip_well_barcode_X',1),
  ('chip_well_barcode_Y',2)
);
COMMIT;THIS IS NOT THE SOLUTION. THIS IS JUST A TEMPORARY RELIEVE BUYING TIME!
Hope this helps !

**Comment by kbergin** [comment]
➤ Atanas Mihalev commented:

Once this issue is said to be “resolved” we need to properly test it. One way to test it is this:

  1. Pipeline-team blacklist a given sample by using ArraysManualBlacklistUpdate tool (this should result in MODIFIED_AT being properly set to current time)
  2. Pipeline-team whitelist the same sample by using ArraysManualBlacklistUpdate tool and hands over the sample-name to Analytics.
  3. 20 minutes later, Analytics checks if that sample has boarded ETL train as expected (we will check if new data has made it into our datamarts)

Notes:

Step 2 can be done immediately after step 1 so no process/system should be affected.

Hope this helps

**Comment by kbergin** [comment]
➤ George Grant commented:

I believe this is being caused by ArraysManualBlacklistUpdate tool being run on prem.   It is currently run in two places.  1) On the cloud, in the Arrays pipeline wdl.  There, it will write to the cloud sql database using UTC.  2) On prem, manually by ops on an as-needed basis.  There it (presumably - need to verify) will use local time.

The fix is (presumably) to update the code to use UTC (both on-prem and in the cloud).

**Comment by kbergin** [comment]
➤ George Grant commented:

[~atanas] can this be tested in a dev environment (we have both a dev cloud sql database).  Testing things in production always makes me nervous.

**Comment by kbergin** [comment]
➤ Atanas Mihalev commented:

I understand. What about this test in dev

  1. Run ArraysManualBlacklistUpdate tool
  2. Shortly after that, check in DB and make sure that MODIFIED_AT is no more than few minutes off SYSDATE()

Following query would do it, just fix where-clause accordingly.

select sysdate(), a.MODIFIED_AT, a.CHIP_WELL_BARCODE
from ARRAYS_QC_BLACKLISTING a
where a.CHIP_WELL_BARCODE='203273230151_R04C01'Run this test against variety of configurations being subject of concerns to achieve good coverage (on prem, in cloud, etc.)

Thanks

**Comment by kbergin** [comment]
➤ Janki Kaneria commented:

Patch went out late evening yesterday. QA will be that we don’t see local timezone in the blacklisting from this morning forward.

**Comment by kbergin** [comment]
➤ Atanas Mihalev commented:

[~jkaneria] Great news - thank you !
Most likely some of blacklisted-before chips have been hurt by this bug and therefore have not been able to board the ETL train. In other words data in Mercury might be out of sync.
Instead of trying to figure out which ones are affected (or worse - just wait for users to complain), it would be much easier just to initiate a full resync - this can be accomplished by merely setting MODIFIED_AT=SYSDATE to all ARRAY_QC_BLACKLISTING records in the cloud.
thanks

cc: [~thompson]
